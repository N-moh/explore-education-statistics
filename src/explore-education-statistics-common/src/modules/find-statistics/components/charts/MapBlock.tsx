/* eslint-disable @typescript-eslint/camelcase */
import {
  Feature,
  FeatureCollection,
  GeoJsonProperties,
  Geometry,
} from 'geojson';
import 'leaflet/dist/leaflet.css';
import React from 'react';
import { GeoJSON, LatLngBounds, Map } from 'react-leaflet';
import { ChartProps } from '@common/modules/find-statistics/components/charts/ChartFunctions';
import { DataBlockGeoJsonProperties } from '@common/services/dataBlockService';
import { FormSelect } from '@common/components/form';
import classNames from 'classnames';
import { SelectOption } from '@common/components/form/FormSelect';
import TimePeriod from '@common/services/types/TimePeriod';
import { Dictionary } from '@common/types/util';
import UKGeoJson from '@common/services/UKGeoJson';

export type MapFeature = Feature<Geometry, GeoJsonProperties>;

interface MapProps extends ChartProps {
  position?: { lat: number; lng: number };
  maxBounds?: LatLngBounds;
}

interface GroupOption {
  id: string;
  title: string;
  type: 'timeperiod' | 'filter' | 'location' | null;
  selected: string;
  options: SelectOption[];
}

function MapBlock(props: MapProps) {
  const {
    position = { lat: 53.00986, lng: -3.2524038 },
    width,
    height,
    meta,
    data,
    dataGroupings = [],
  } = props;

  const locationGeoJSON = Object.values(meta.locations)
    .map(({ geoJson }) => geoJson)
    .filter(gj => gj !== undefined)
    .map(geojsonArray => geojsonArray[0]);

  // Build options for forms
  const locationOptions = [
    { label: 'All', value: '' },

    ...locationGeoJSON
      .map(({ properties: { name, code } }) => ({ label: name, value: code }))
      .sort((a, b) => {
        if (a.label < b.label) return -1;
        if (a.label > b.label) return 1;
        return 0;
      }),
  ];

  const groupOptions: GroupOption[] = dataGroupings.map(
    ({ timePeriod }, index) => {
      if (timePeriod) {
        return {
          id: `groupOption_year_${index}`,
          title: 'Select a time period',
          type: 'timeperiod',
          selected: `${data.result[0].year}`,
          options: Object.values<SelectOption>(
            data.result.reduce(
              (result: Dictionary<TimePeriod>, { year, timeIdentifier }) => {
                return {
                  ...result,
                  [`${year}${timeIdentifier}`]: new TimePeriod(
                    year,
                    timeIdentifier,
                  ),
                };
              },
              {},
            ),
          )
            .sort()
            .reverse(),
        };
      }

      return {
        id: `groupOption_${index}`,
        title: '',
        selected: '',
        type: null,
        options: Array<SelectOption>(),
      };
    },
  );

  const mapRef = React.createRef<Map>();

  const getGeometryForOptions = () => {
    const value: FeatureCollection<Geometry, DataBlockGeoJsonProperties> = {
      type: 'FeatureCollection',
      features: locationGeoJSON,
    };

    return value;
  };

  const [geometry, updateGeometry] = React.useState(getGeometryForOptions());

  const updateGeometryForOptions = () => {
    updateGeometry(getGeometryForOptions());
  };

  requestAnimationFrame(() => {
    if (mapRef.current) {
      mapRef.current.leafletElement.invalidateSize();
    }
  });

  const [selectedLocation, updateSelectedLocation] = React.useState('');

  const selectLocationOption = (locationValue: string) => {
    updateSelectedLocation(locationValue);
    // updateGeometryForOptions();
  };

  const updateSelectedGroupOption = (group: GroupOption, value: string) => {
    // eslint-disable-next-line no-param-reassign
    group.selected = value;
    updateGeometryForOptions();
  };

  const onEachFeature = (feature: MapFeature) => {
    // eslint-disable-next-line
    console.log(feature);
  };

  const uk = UKGeoJson.UK;

  return (
    <div className="govuk-grid-row">
      <div
        className={classNames('govuk-grid-column-one-third')}
        aria-live="assertive"
      >
        <form>
          <div className="govuk-form-group govuk-!-margin-bottom-6">
            <FormSelect
              name="selectedLocation"
              id="selectedLocation"
              label="Select a location"
              value={selectedLocation}
              onChange={e => selectLocationOption(e.currentTarget.value)}
              order={null}
              options={locationOptions}
            />
          </div>

          {groupOptions.map(grouping => (
            <div
              key={grouping.id}
              className="govuk-form-group govuk-!-margin-bottom-6"
            >
              <FormSelect
                name="grouping"
                id={grouping.id}
                label={grouping.title}
                value={grouping.selected}
                onChange={e =>
                  updateSelectedGroupOption(grouping, e.currentTarget.value)
                }
                order={null}
                options={grouping.options}
              />
            </div>
          ))}
        </form>

        {/*
        <div>
          <div className="dfe-dash-tiles__tile govuk-!-margin-bottom-6">
            <h3 className="govuk-heading-m dfe-dash-tiles__heading">
              Overall absence
            </h3>
            <p
              className="govuk-heading-xl govuk-!-margin-bottom-2"
              aria-label="Overall absence"
            >
              <span> ---% </span>
            </p>
            <Details summary="What is overall absence?">
              Overall absence is the adipisicing elit. Dolorum hic nobis
              voluptas quidem fugiat enim ipsa reprehenderit nulla.
            </Details>
          </div>
          <div className="dfe-dash-tiles__tile govuk-!-margin-bottom-6">
            <h3 className="govuk-heading-m dfe-dash-tiles__heading">
              Authorised absence
            </h3>
            <p
              className="govuk-heading-xl govuk-!-margin-bottom-2"
              aria-label="Authorised absence"
            >
              <span> ---% </span>
            </p>
            <Details summary="What is authorised absence?">
              Authorised absence is the adipisicing elit. Dolorum hic nobis
              voluptas quidem fugiat enim ipsa reprehenderit nulla.
            </Details>
          </div>
          <div className="dfe-dash-tiles__tile govuk-!-margin-bottom-6">
            <h3 className="govuk-heading-m dfe-dash-tiles__heading">
              Unauthorised absence
            </h3>
            <p
              className="govuk-heading-xl govuk-!-margin-bottom-2"
              aria-label="Unauthorised absence"
            >
              <span> ---% </span>
            </p>
            <Details summary="What is unauthorised absence?">
              Unauthorised absence is the adipisicing elit. Dolorum hic
              nobis voluptas quidem fugiat enim ipsa reprehenderit nulla.
            </Details>
          </div>
        </div>
        */}

        {/*
        <div className={classNames('')}>
          <h3 className="govuk-heading-s">Key to overall absence rate</h3>
          <dl className="govuk-list">
            {legend &&
            legend.map(({min, max, idx}) => (
              <dd key={idx}>
                <span className={styles[`rate${idx}`]}>&nbsp;</span> {min}%
                to {max}%{' '}
              </dd>
            ))}
          </dl>
        </div>
        */}
      </div>

      <div className={classNames('govuk-grid-column-two-thirds')}>
        <Map
          style={{
            width: (width && `${width}px`) || '100%',
            height: `${height || 600}px`,
          }}
          ref={mapRef}
          center={position}
          // className={styles.map}
          zoom={6.5}
          // minZoom={6.5}
          // zoomSnap={0.5}
          // maxBounds={this.state.maxBounds}
        >
          <GeoJSON
            data={geometry}
            onEachFeature={onEachFeature}
            // style={this.styleFeature}
            // onClick={this.handleClick}
          />
          {uk ? <GeoJSON data={uk} style={{ color: '777', weight: 1 }} /> : ''}
        </Map>
      </div>
    </div>
  );
}

export default MapBlock;
