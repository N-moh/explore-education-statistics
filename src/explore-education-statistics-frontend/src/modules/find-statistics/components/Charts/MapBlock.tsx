import {
  Feature,
  FeatureCollection,
  GeoJsonProperties,
  Geometry,
  GeometryCollection,
} from 'geojson';
import { LatLngBounds } from 'leaflet';
import 'leaflet/dist/leaflet.css';
import React, { Component } from 'react';
import { GeoJSON, Map } from 'react-leaflet';

import { ChartProps } from './Charts';

interface MapFeature extends Feature<Geometry, GeoJsonProperties> {}

interface MapState {
  position: { lat: number; lng: number };
  maxBounds: LatLngBounds;
  geometry: MapFeature;
}

interface MapProps extends ChartProps {
  geometry?: MapFeature;
}

export class MapBlock extends Component<MapProps, MapState> {
  protected mapRef: Map | null;

  public state = {
    maxBounds: new LatLngBounds({ lat: 48, lng: -6.5 }, { lat: 60, lng: 2 }),

    position: {
      lat: 53.009865,
      lng: -3.2524038,
    },
  } as any;

  constructor(props: MapProps) {
    super(props);
    this.mapRef = null;

    if (props.geometry !== undefined) {
      // @ts-ignore
      this.state.geometry = props.geometry;
    }
  }

  private renderGeoJson() {
    if (this.state.geometry === undefined) return '';

    return (
      <GeoJSON
        data={this.state.geometry}
        // onEachFeature={this.onEachFeature}
        // style={this.styleFeature}
        // onClick={this.handleClick}
      />
    );
  }

  public refresh() {
    requestAnimationFrame(_ => {
      if (this.mapRef) {
        this.mapRef.leafletElement.invalidateSize();
      }
    });
  }

  public componentDidMount() {
    // force a refresh to fix a bug
    requestAnimationFrame(() => this.refresh());
  }

  public render() {
    return (
      <div>
        <Map
          style={{ height: `${this.props.height || 200}px` }}
          ref={el => (this.mapRef = el)}
          center={this.state.position}
          // className={styles.map}
          zoom={6.5}
          // minZoom={6.5}
          // zoomSnap={0.5}
          // maxBounds={this.state.maxBounds}
        >
          {this.renderGeoJson()}
        </Map>
      </div>
    );
  }
}

export default MapBlock;
