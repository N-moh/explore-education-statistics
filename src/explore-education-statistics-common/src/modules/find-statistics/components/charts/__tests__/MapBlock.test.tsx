/* eslint-disable */
import React from 'react';

import { render, wait } from 'react-testing-library';

import testData from '@common/modules/find-statistics/components/charts/__tests__/__data__/testBlockData';
import MapBlock from '../MapBlock';

describe('MapBlock', () => {
  test('renders', async () => {
    const { container } = render(
      <MapBlock
        {...{
          ...testData.AbstractMultipleChartProps,
          axes: {
            ...testData.AbstractMultipleChartProps.axes,
            major: {
              ...testData.AbstractMultipleChartProps.axes.major,
              groupBy: 'locations',
            },
          },
        }}
        height={600}
        width={900}
      />,
    );

    await wait();

    expect(container).toMatchSnapshot();
  });

  test('includes all locations in select', async () => {
    const { container } = render(
      <MapBlock
        {...{
          ...testData.AbstractMultipleChartProps,
          axes: {
            ...testData.AbstractMultipleChartProps.axes,
            major: {
              ...testData.AbstractMultipleChartProps.axes.major,
              groupBy: 'locations',
            },
          },
        }}
        height={600}
        width={900}
      />,
    );

    await wait();

    const select = container.querySelector('#selectedLocation');

    expect(select).toBeVisible();

    if (select) {
      expect(
        select.querySelector('option[value="E92000001"]'),
      ).toHaveTextContent('England');
      expect(
        select.querySelector('option[value="S92000001"]'),
      ).toHaveTextContent('Scotland');
    }
  });

  test('includes all indicators in select', async () => {
    const { container } = render(
      <MapBlock
        {...{
          ...testData.AbstractMultipleChartProps,
          axes: {
            ...testData.AbstractMultipleChartProps.axes,
            major: {
              ...testData.AbstractMultipleChartProps.axes.major,
              groupBy: 'locations',
            },
          },
        }}
        height={600}
        width={900}
      />,
    );

    await wait();

    const select = container.querySelector('#selectedIndicator');

    expect(select).toBeVisible();

    if (select) {
      expect(select.querySelector('option[value="2"]')).toHaveTextContent(
        'Authorised absence rate',
      );
      expect(select.querySelector('option[value="1"]')).toHaveTextContent(
        'Overall absence rate',
      );
      expect(select.querySelector('option[value="0"]')).toHaveTextContent(
        'Unauthorised absence rate',
      );
    }
  });

  test('include all indicators from reduced selection', async () => {
    const { container } = render(
      <MapBlock
        {...{
          ...testData.AbstractLargeDataChartProps_smaller_datasets,
          axes: {
            ...testData.AbstractLargeDataChartProps_smaller_datasets.axes,
            major: {
              ...testData.AbstractLargeDataChartProps_smaller_datasets.axes
                .major,
              groupBy: 'locations',
            },
          },
        }}
        height={600}
        width={900}
      />,
    );

    await wait();

    const select = container.querySelector('#selectedIndicator');

    expect(select).toBeVisible();

    if (select) {
      expect(select.querySelector('option[value="1"]')).toHaveTextContent(
        'Overall absence rate',
      );
      expect(select.querySelector('option[value="0"]')).toHaveTextContent(
        'Unauthorised absence rate',
      );
    }
  });
});
