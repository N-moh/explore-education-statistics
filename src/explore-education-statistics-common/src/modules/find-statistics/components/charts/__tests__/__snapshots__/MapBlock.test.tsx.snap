// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapBlock MapBlock renders 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      aria-live="assertive"
      class="govuk-grid-column-one-third"
    >
      <form>
        <div
          class="govuk-form-group govuk-!-margin-bottom-6"
        >
          <label
            class="govuk-label"
            for="selectedLocation"
          >
            Select a location
          </label>
          <select
            class="govuk-select select"
            id="selectedLocation"
            name="selectedLocation"
          >
            <option
              value=""
            >
              All
            </option>
            <option
              value="E92000001"
            >
              England
            </option>
          </select>
        </div>
      </form>
    </div>
    <div
      class="govuk-grid-column-two-thirds"
    >
      <div
        class="leaflet-container leaflet-grab leaflet-touch-drag"
        style="width: 900px; height: 600px; position: relative;"
        tabindex="0"
      >
        <div
          class="leaflet-pane leaflet-map-pane"
          style="left: 0px; top: 0px;"
        >
          <div
            class="leaflet-pane leaflet-tile-pane"
          />
          <div
            class="leaflet-pane leaflet-shadow-pane"
          />
          <div
            class="leaflet-pane leaflet-overlay-pane"
          >
            <svg
              height="0"
              pointer-events="none"
              style="left: 0px; top: 0px;"
              viewBox="0 0 0 0"
              width="0"
            >
              <g>
                <path
                  class="leaflet-interactive"
                  d="M0 0"
                  fill="#3388ff"
                  fill-opacity="0.2"
                  fill-rule="evenodd"
                  stroke="#3388ff"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-opacity="1"
                  stroke-width="3"
                />
                <path
                  class="leaflet-interactive"
                  d="M-1 1L-1 -1L1 -1L1 1z"
                  fill="777"
                  fill-opacity="0.2"
                  fill-rule="evenodd"
                  stroke="777"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-opacity="1"
                  stroke-width="1"
                />
              </g>
            </svg>
          </div>
          <div
            class="leaflet-pane leaflet-marker-pane"
          />
          <div
            class="leaflet-pane leaflet-tooltip-pane"
          />
          <div
            class="leaflet-pane leaflet-popup-pane"
          />
        </div>
        <div
          class="leaflet-control-container"
        >
          <div
            class="leaflet-top leaflet-left"
          >
            <div
              class="leaflet-control-zoom leaflet-bar leaflet-control"
            >
              <a
                aria-label="Zoom in"
                class="leaflet-control-zoom-in"
                href="#"
                role="button"
                title="Zoom in"
              >
                +
              </a>
              <a
                aria-label="Zoom out"
                class="leaflet-control-zoom-out"
                href="#"
                role="button"
                title="Zoom out"
              >
                âˆ’
              </a>
            </div>
          </div>
          <div
            class="leaflet-top leaflet-right"
          />
          <div
            class="leaflet-bottom leaflet-left"
          />
          <div
            class="leaflet-bottom leaflet-right"
          >
            <div
              class="leaflet-control-attribution leaflet-control"
            >
              <a
                href="https://leafletjs.com"
                title="A JS library for interactive maps"
              >
                Leaflet
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
